File: conv_with_pad/conv3d_nchw.cpp
#include <iostream>
#include <vector>
#include "../utils/utils.hpp"

using namespace std;

// Define the convolution function
void conv3d(const vector<vector<vector<vector<float>>>> &input,
            const vector<vector<vector<vector<float>>>> &kernel,
            vector<vector<vector<vector<float>>>> &output,
            int stride = 1, int padding = 1) {

    int batch_size = input.size();
    int in_channels = input[0].size();
    int out_channels = kernel.size();
    int input_height = input[0][0].size();
    int input_width = input[0][0][0].size();
    int kernel_height = kernel[0][0].size();
    int kernel_width = kernel[0][0][0].size();

    int output_height = (input_height - kernel_height + 2 * padding) / stride + 1;
    int output_width = (input_width - kernel_width + 2 * padding) / stride + 1;

    // Initialize the output tensor
    output.assign(batch_size, vector<vector<vector<float>>>(
                                  out_channels, vector<vector<float>>(
                                                  output_height, vector<float>(output_width, 0))));

    vector<vector<vector<vector<float>>>> padded_input(batch_size, vector<vector<vector<float>>>(
            in_channels, vector<vector<float>>(
            input_height + 2 * padding, vector<float>(input_width + 2 * padding, 0))));

    for (int b = 0; b < batch_size; ++b) {
        for (int c = 0; c < in_channels; ++c) {
            for (int i = 0; i < input_height; ++i) {
                for (int j = 0; j < input_width; ++j) {
                    padded_input[b][c][i + padding][j + padding] = input[b][c][i][j];
                }
            }
        }
    }
    // // Perform the convolution operation
    for (int b = 0; b < batch_size; ++b) {
        for (int oc = 0; oc < out_channels; ++oc) {
            for (int i = 0; i < output_height; ++i) {
                for (int j = 0; j < output_width; ++j) {
                    float sum = 0.0;
                    for (int ic = 0; ic < in_channels; ++ic) {
                        for (int ki = 0; ki < kernel_height; ++ki) {
                            for (int kj = 0; kj < kernel_width; ++kj) {
                                int input_i = i * stride + ki;
                                int input_j = j * stride + kj;
                                sum += padded_input[b][ic][input_i][input_j] * kernel[oc][ic][ki][kj];
                            }
                        }
                    }
                    output[b][oc][i][j] = sum;
                }
            }
        }
    }
    vector<float> output_1d;
    for (int b = 0; b < batch_size; b++)
        for (int i = 0; i < out_channels; i++)
        {
            for (int j = 0; j < output_height; j++)
            {
                for (int k = 0; k < output_width; k++)
                {
                    output_1d.push_back(output[b][i][j][k]);
                }
            }
        }

    write_to_binary("../outputs/conv3d_nchw_cpp.bin", output_1d);
}

int main() {
    vector<int> input_dims = {1,3,224,224};
    vector<int> kernel_dims = {64,3,7,7};
    auto input = read_npy_file("../inputs/py_input.npy", input_dims);
    auto kernel = read_npy_file("../weights/conv1_wt.npy", kernel_dims);
    // Output tensor
    vector<vector<vector<vector<float>>>> output;

    // // // Perform the convolution
    conv3d(input, kernel, output, 1, 1);

    // // Print the output dimensions and the first value for verification
    cout << "Output shape: [" << output.size() << ", "
         << output[0].size() << ", "
         << output[0][0].size() << ", "
         << output[0][0][0].size() << "]" << endl;
    cout << "First output value: " << output[0][0][0][0] << endl;

    return 0;
}

File: python/conv3d_nchw.py
from torch import nn
import torch
import numpy as np

def write_binary_file(file_path, array):
    with open(file_path, 'wb') as f:
        array.tofile(f)

def write_npy_file(file_path, array):
    np.save(file_path, array)



if __name__ == "__main__":
    input_dims = [1,3,224,224]
    kernel_dims = [64,3,7,7]

    input_height = input_dims[2]
    input_width = input_dims[3]
    input_channel = input_dims[1]

    kernel_height = kernel_dims[2]
    kernel_width = kernel_dims[3]
    kernel_channel = kernel_dims[1]

    stride = 1
    padding = 1
    output_channel = kernel_dims[0]

    output_height = int((input_height + 2 * padding - kernel_height) / stride + 1)
    output_width = int((input_width + 2 * padding - kernel_width) / stride + 1)
    output_channel = kernel_dims[0]

    conv = nn.Conv2d(
        output_channel,
        input_channel,
        kernel_size=(kernel_height, kernel_height),
        stride=stride,
        padding=padding,
    )
    input_data = np.load("../inputs/py_input.npy")
    input_tensor = torch.from_numpy(input_data)

    kernel_data = np.load("../weights/conv1_wt.npy")
    kernel_tensor = torch.from_numpy(kernel_data)

    temp = 0


    bias_matrix = [(0) for i in range(output_channel)]
    bias_tensor = torch.Tensor(bias_matrix)

    conv.weight.data = kernel_tensor
    conv.bias.data = bias_tensor

    output = conv(input_tensor)
    print("input shape ", input_tensor.shape)
    print("kernel shape ", kernel_tensor.shape)
    print("output shape ", output.shape)
    output_np = output.detach().numpy() 
    output_1d = output_np.flatten()    # Flatten to 1D array

    # Write the 1D array to a binary file
    write_binary_file('../outputs/py_conv3d_nchw_output.bin', output_1d)

    print("3D Convolution complete. Output written to 'py_filter_output.bin'.")